selfdrive/car/hyundai/interface.py

```diff
@@ -0,0 +1,259 @@
+
+
+    # no rear steering, at least on the listed cars above
+    ret.steerRatioRear = 0.
+    ret.steerControlType = car.CarParams.SteerControlType.torque
+
+    # steer, gas, brake limitations VS speed
+    ret.steerMaxBP = [0.]
+    ret.steerMaxV = [1.0]
+    ret.gasMaxBP = [0.]
+    ret.gasMaxV = [1.]
+    ret.brakeMaxBP = [0.]
+    ret.brakeMaxV = [1.]
+    ret.longPidDeadzoneBP = [0.]
+    ret.longPidDeadzoneV = [0.]
+
+    ret.enableCamera = not any(x for x in CAMERA_MSGS if x in fingerprint)
+
+    ret.steerLimitAlert = False
+    ret.stoppingControl = False
+    ret.startAccel = 0.0
+
+    return ret
+
+  # returns a car.CarState
+  def update(self, c):
+    # ******************* do can recv *******************
+    canMonoTimes = []
+    self.cp.update(int(sec_since_boot() * 1e9), False)
+    self.cp_cam.update(int(sec_since_boot() * 1e9), False)
+    self.CS.update(self.cp, self.cp_cam)
+    # create message
+    ret = car.CarState.new_message()
+    # speeds
+    ret.vEgo = self.CS.v_ego
+    ret.vEgoRaw = self.CS.v_ego_raw
+    ret.aEgo = self.CS.a_ego
+    ret.yawRate = self.CS.yaw_rate
+    ret.standstill = self.CS.standstill
+    ret.wheelSpeeds.fl = self.CS.v_wheel_fl
+    ret.wheelSpeeds.fr = self.CS.v_wheel_fr
+    ret.wheelSpeeds.rl = self.CS.v_wheel_rl
+    ret.wheelSpeeds.rr = self.CS.v_wheel_rr
+
+    # gear shifter
+    ret.gearShifter = self.CS.gear_shifter
+
+    # gas pedal
+    ret.gas = self.CS.car_gas
+    ret.gasPressed = self.CS.pedal_gas > 1e-3   # tolerance to avoid false press reading
+
+    # brake pedal
+    ret.brake = self.CS.user_brake
+    ret.brakePressed = self.CS.brake_pressed != 0
+    ret.brakeLights = self.CS.brake_lights
+
+    # steering wheel
+    ret.steeringAngle = self.CS.angle_steers
+    ret.steeringRate = self.CS.angle_steers_rate  # it's unsigned
+
+    ret.steeringTorque = self.CS.steer_torque_driver
+    ret.steeringPressed = self.CS.steer_override
+
+    # cruise state
+    ret.cruiseState.enabled = self.CS.pcm_acc_status != 0
+    if self.CS.pcm_acc_status != 0:
+      ret.cruiseState.speed = self.CS.cruise_set_speed
+    else:
+      ret.cruiseState.speed = 0
+    ret.cruiseState.available = bool(self.CS.main_on)
+    ret.cruiseState.standstill = False
+
+    # TODO: button presses
+    buttonEvents = []
+
+    if self.CS.left_blinker_on != self.CS.prev_left_blinker_on:
+      be = car.CarState.ButtonEvent.new_message()
+      be.type = 'leftBlinker'
+      be.pressed = self.CS.left_blinker_on != 0
+      buttonEvents.append(be)
+
+    if self.CS.right_blinker_on != self.CS.prev_right_blinker_on:
+      be = car.CarState.ButtonEvent.new_message()
+      be.type = 'rightBlinker'
+      be.pressed = self.CS.right_blinker_on != 0
+      buttonEvents.append(be)
+
+    ret.buttonEvents = buttonEvents
+    ret.leftBlinker = bool(self.CS.left_blinker_on)
+    ret.rightBlinker = bool(self.CS.right_blinker_on)
+
```

Comment:
This is not sufficient, the flash is required.
The car can be configured to 3, 5, 7, or 9 flashes on momentary push.
only while the switch is in the momentary position does the blinker signal come on.
As such, you must monitor the flash.

Reply:
makes sense